Domain Upsell Back Button: Previous Page or Preferred View

chad1008 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

7639-gh-Automattic/dotcom-forge

## Proposed Changes

Have the "Home" link on the domain with plan upsell page take you to either wp-admin or My Home, depending on your current site admin interface preferences. If this value isn't available, we'll keep using the most recent page from browser history as a backup.

The issue linked above recommends using browser history first and My Home/wp-admin as the fallback, but we've shifted away from that for the reasons [discussed in the comments](https://github.com/Automattic/wp-calypso/pull/91957#discussion_r1654574140).

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Sending a site owner who prefers wp-admin to My Home could be jarring, and not the best experience, since they've expressed that this isn't how the prefer to interact with their site.

## Testing Instructions

1. Open WordPress.com and select a site free site with no plan or domain name.
2. Make note of your test site's ID
3. Simple sites don't currently have access to the site admin interface setting, but there is work underway (see: pfsHM7-EF-p2) to change that. You can activate the upcoming setting on your test site by ssh'ing into your sandbox and running:
	- `wpsh`
	- `update_blog_option( blog_id, 'wpcom_admin_interface', 'wp-admin' );`
	- `update_blog_option( blog_id, 'wpcom_classic_early_release', 1 );`
4. Locate the "Free domain with annual plan" nudge and click the "upgrade" button
5. Confirm that the "Home" button takes you to your test site's wp-admin (note: this assumes you ran the above commands to change your site's interface preference to wp-admin).
6. To switch back to the default/calypso interface, run the following command via `wpsh` on your sandbox:
	- `update_blog_option( blog_id, 'wpcom_admin_interface', 'default' );`
7. Visit your test site's My Home page locally at `calypso.localhost:3000/home/[SITE_URL]. Make sure you manually get yourself back to your local Calypso, because that wp-admin link in the previous step will have taken you to production!
8. Once you've loaded your My Home page, test the Upgrade button and the Home link again
9. Confirm that this time, with the interface setting changed, the button takes you back to My Home, and not wp-admin

<details>
<summary>10. To simulate a scenario where our desired "home" link isn't available, apply the following diff:</summary>

```diff
diff --git a/client/my-sites/domains/domain-search/index.jsx b/client/my-sites/domains/domain-search/index.jsx
index 12b422f4ab..b72ea4dfb6 100644
--- a/client/my-sites/domains/domain-search/index.jsx
+++ b/client/my-sites/domains/domain-search/index.jsx
@@ -319,10 +319,7 @@ class DomainSearch extends Component {
 							goBackLink: `/setup/${ siteIntent }/launchpad?siteSlug=${ selectedSiteSlug }`,
 					  }
 					: {
-							goBackLink:
-								this.props.preferredView === 'wp-admin'
-									? this.props.wpAdminUrl
-									: `/home/${ selectedSiteSlug }`,
+							goBackLink: '',
 					  };
 			content = (
 				<span>

```

</details>

11. Navigate to a different area of Calypso (ie, Pages, or Posts) and test the Upgrade button and Home links again. Confirm that they return you to your most recent page.

chad1008 committed:
update home/back button label and logic

fredrikekelund commented on client/my-sites/domains/components/domain-and-plan-package/navigation/index.jsx:
```diff
@@ -7,10 +7,13 @@ export default function DomainAndPlanPackageNavigation( props ) {
 	const goBack = () => {
+		if ( window.history.length > 1 ) {
+			window.history.go( -1 );
+			return;
+		}
+
 		if ( props.goBackLink ) {
 			window.location.assign( props.goBackLink );
-		} else {
-			window.history.go( -1 );
 		}
```
This change _seems_ to work, but `history.length` isn't reduced after `history.go( -1 )`, which means that if a user reaches the `Select a plan` step, we'll always run this code instead of `location.assign()`.

Maybe it's safer to revert this specific change..?

chad1008 replied:
I'm comfortable with that. The motivation for this change was the 
> It should link to the previous page if possible. If not, go to `:site/wp-admin` if the site is configured to the classic interface. Otherwise, go to `/home/:site`

portion of the [related issue](https://href.li/?7639-gh-Automattic/dotcom-forge)'s acceptance criteria, but it did feel a little odd to me that we'd virtually never, or just never never, actually make it to the `location.assign` flow.

If we revert this change, I'd probably also opt to revert the change to the button text, and make it say **Home** by default, and conditionally say **Back** if that path is utilized (which is what it does currently on production)

Effectively, this PR would really just make sure the **Home** path is conditional based on the user's interface preference, but leave all other behavior as it is.

That all works for me. Let me know if you have any objections/concerns @arthur791004 🙂

fredrikekelund commented on client/my-sites/domains/domain-search/index.jsx:
```diff
@@ -444,6 +449,8 @@ export default connect(
 				isSiteOnWooExpress( state, siteId ) ||
 				isSiteOnEcommerce( state, siteId ),
 			isFromMyHome: getCurrentQueryArguments( state )?.from === 'my-home',
+			preferredView: getPreferredEditorView( state, siteId ),
```
`getPreferredEditorView` seems to only be used in the context of the site editor. I don't know that it's wrong to use here, but I want to double check if you've looked at other candidates, @chad1008?

What we're looking for here is the value of this setting, right..? cc @arthur791004 

I guess we might be introducing a similar setting for Simple sites with this project pfsHM7-EF-p2

<img width="740" alt="Screenshot 2024-06-26 at 12 42 17" src="https://github.com/Automattic/wp-calypso/assets/1101677/ce0ea3bd-4399-48ff-b350-813117c72263">

arthur791004 replied:
The value can be retrieved by `getSiteOption( state, site?.ID, 'wpcom_admin_interface' );`
* Classic style - the value equals to `wp-admin`

chad1008 replied:
Thanks @fredrikekelund and @arthur791004. I'll update to PR to use this this method instead.

chad1008 replied:
Following up on this feedback, I think I may have started work on a free site, and then for whatever reason completed an upgrade, which gave me access to the **Settings > General > Admin Interface** that isn't normally available on free sites.

As @fredrikekelund noted above, that option is coming for simple sites, so I've added the `wpsh` command to enable that setting to the testing instructions above.

fredrikekelund reviewed (commented):
Some drive by comments 🚗

I was only able to find the `Free domain with an annual plan` upsell for a site on the free plan. For a new site on an annual Creator plan without a domain, I get a different upsell (`Free domain available`). Do you know when this upsell is activated, @chad1008?

chad1008 commented:
> I was only able to find the `Free domain with an annual plan` upsell for a site on the free plan. For a new site on an annual Creator plan without a domain, I get a different upsell (`Free domain available`). Do you know when this upsell is activated, @chad1008?

The upsell we're working on (Free domain with an annual plan) is only triggered for free sites (you're correct, there. I've updated my testing instructions to be more clear on that front!).

The other upsell (Free domain available) appears to by triggered by any site with an available domain credit, but it goes to a different flow that this PR won't touch.

chad1008 committed:
feedback: prefer `getSiteOption` over `getPreferredEditorView`

chad1008 commented:
Update: I've addressed the `getSiteOption` vs `getPreferredEditorView` feedback. I'll take care of the other bit regarding the back behavior overshadowing the "home" behavior early next week, but I wanted to give time for @arthur791004 to share any thoughts first.

chad1008 committed:
addressing feedback: link to home/wp-admin first, browser history second

chad1008 commented:
I've updated the PR to reflect the above conversation:

- Link will say "Home" or "Back" depending on the situation
- If possible, the link will take you to either wp-admin or My Home, depending on your current admin interface preference. In this case the link text will be "Home"
- If needed, the link will revert to reading "Back" and will take you to your previous page based on browser history

I've also updated the pr description and testing instructions to match these changes. @nightnei, would you mind looking this one over and sharing any thoughts/approving? We've shifted a bit from the original plan laid out by @arthur791004, so I'd love some extra eyes before we merge 😄

nightnei commented on client/my-sites/domains/components/domain-and-plan-package/navigation/index.jsx:
```diff
@@ -6,10 +6,16 @@ import './style.scss';
+			return;
+		}
+
+		if ( window.history.length > 1 ) {
```
You added the condition; hence, does it mean that we have a case when history length is <= 1? And in this case clicking on the button do nothing. We should definitely handle default behaviour. Or if it's not possible to have length less than 1, then let's remove the condition, since it's redundant and can just confuse developers in the future.

nightnei replied:
Also, as I see, `window.history.go( -1 );` was redundant before, since we always had `goBackLink`, according to the old condition:

```
const goBackButtonProps = launchpadScreen === 'full'
	? {
			goBackText: translate( 'Next Steps' ),
			goBackLink: `/setup/${ siteIntent }/launchpad?siteSlug=${ selectedSiteSlug }`,
	  }
	: {
			goBackLink: `/home/${ selectedSiteSlug }`,
	  };
```

So, maybe we can just remove it?

UPDATE: now you added `this.props.preferredView === 'wp-admin`, so looks like, only now, [it can be undefined](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/sites/selectors/get-site-admin-url.ts#L20). But can it be?
AFAIU, if there is `this.props.preferredView`, then we should definitely have `getSiteOption( state, siteId, 'admin_url' )` from `getSiteAdminUrl`, or I understand incorrectly?
Could you please check it, since I suppose that we can just delete it. And avoid having code just to have code which does nothing :)

chad1008 replied:
Thanks for these points @nightnei. You're right that with the original logic, the `window.history.go(-1)` was redundant, because `goBackLink` was always populated by a string. As long as that holds true, after addressing the rest of your feedback, I agree it can be removed.

Looking at the new logic, you're also right that `wpAdminUrl` could be `null`. I think the current change is still mostly safe, because (as you pointed out) if we're able to declare `preferredView` as `wp-admin`,  then we should have no trouble retrieving the `wpAdminUrl`. Both of those selectors are relying on `siteId`, so if one succeeds, they should both succeed. If `preferredView` WAS `null`, we'd skip to the second path of the ternary, and the having `wpAdminUrl` be `null` doesn't matter. We'd be sending them to Calypso at that point.

But I hate that I used the word "should" just now. I don't see a scenario where `preferredView === 'wp-admin'` but `wpAdminUrl === null`, but better safe than sorry. I'm going to add a check of the `wpAdminUrl` value into that ternary so we can be sure it's populated before using for the `goBackLink` value.

chad1008 committed:
add safety check of `wpAdminUrl`

chad1008 committed:
remove redundant logic

nightnei commented on client/my-sites/domains/components/domain-and-plan-package/navigation/index.jsx:
```diff
@@ -6,11 +6,12 @@ import './style.scss';
 
+	// `goBackLink` will be either wp-admin or My Home, depending on the user's
+	// current admin interface preference.
+	// If unavailable, we fall back to the previous page in history.
```
Now we don't need it
```suggestion
```

nightnei commented on client/my-sites/domains/domain-search/index.jsx:
```diff
@@ -323,9 +324,11 @@ class DomainSearch extends Component {
 					: {
-							goBackLink: `/home/${ selectedSiteSlug }`,
+							goBackLink:
+								this.props.preferredView === 'wp-admin' && !! this.props.wpAdminUrl
```
Awesome 👍

nightnei commented on client/my-sites/domains/components/domain-and-plan-package/navigation/index.jsx:
```diff
@@ -6,11 +6,12 @@ import './style.scss';
+	// current admin interface preference.
+	// If unavailable, we fall back to the previous page in history.
 	const goBack = () => {
 		if ( props.goBackLink ) {
```
Also, it seems, now, we can remove this condition too
```suggestion
```

fredrikekelund replied:
If we expect `goBackLink` to always be defined in the props, then we should enforce it. I suggest we do that by converting this file to TS 👍

chad1008 committed:
Update client/my-sites/domains/components/domain-and-plan-package/navigation/index.jsx

Co-authored-by: Volodymyr Makukha <nei.css@gmail.com>

chad1008 committed:
remove now-obsolete condition

chad1008 committed:
convert DomainAndPlanPackageNavigation to TS

chad1008 commented on client/my-sites/domains/components/domain-and-plan-package/navigation/index.tsx:
```diff
@@ -26,7 +29,7 @@ export default function DomainAndPlanPackageNavigation( props ) {
 		<div className="domain-and-plan-package-navigation">
 			<div className="domain-and-plan-package-navigation__back">
-				<Button borderless="true" onClick={ goBack }>
+				<Button borderless onClick={ goBack }>
```
After converting the file, TS flagged this boolean-as-string. After fixing, the linter helpfully pointed out that the value should be omitted. I've tested locally to confirm the button style isn't impacted by this change.

chad1008 committed:
convert DomainSearch to typescript

chad1008 commented:
Worked through most of the TS conversion work before I ran out of time tonight. Still a WIP, as there are a couple of pesky errors left, and It likely needs some refinement. I got a little heavy handed in some points and there are probably more elegant solutions to some of these errors. Feedback is welcome if either of you feel like taking a look @fredrikekelund/@nightnei

chad1008 commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -53,62 +53,94 @@ import {
-		} );
-		this.isMounted && page( domainMappingUrl );
+	handleAddMapping = ( domain: string ) => {
+		if ( this.props.selectedSiteSlug ) {
```
This check ensures `selectedSiteSlug` is not `null` before proceeding. I'm not thrilled with this approach, and would love some feedback.  I'd like to be able to more confidently assert that it will never be `null` (rather than just appeasing TS as I'm doing now), or add something to ensure a `null` value is handled gracefully.

nightnei replied:
I see :( 

Let's go the next way:

```
// Just TS typing fix, we always have selectedSiteSlug
if ( !this.props.selectedSiteSlug ) {
	return;
}

.... rest of teh function
```

Cooment will be very useful, in case if in the future somebody does refactoring. Also it will help to understand the purpose, if another Code Wrangler will do maintenance of this function

chad1008 commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -53,62 +53,94 @@ import {
 		try {
-			await this.props.shoppingCartManager.addProductsToCart( [ domainTransfer( { domain } ) ] );
+			await this.props.shoppingCartManager.addProductsToCart( [
+				domainTransfer( { domain, extra: {} } ),
```
This empty object works, but I don't really like it.

`domainTransfer` requires an `extra` prop, but [all of the properties of that interface are optional](https://github.com/Automattic/wp-calypso/blob/c2c6ec18d29fa98a36b771c68b2bba92a4ee7b00/packages/shopping-cart/src/types.ts#L649-L680). I'm tempted to make the `extra` param of this function optional itself, rather than being forced to pass this empty object. Thoughts?

nightnei replied:
I belive that we can safely update `domainTransfer` function, by setting `extra` as optional value. It should not make any issues in the code. And additionally the function expects extra as optional, since inside we have ternary operator `...( properties.extra ? { extra: properties.extra } : {} ),`, so let's proceed with updating `domainTransfer` function and remove this empty object `{}`.

chad1008 commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -173,7 +204,7 @@ class DomainSearch extends Component {
 		let registration = domainRegistration( {
 			domain,
-			productSlug,
+			productSlug: productSlug as string,
```
I dislike this immensely, but drew a blank a better approach than just heavy-handedly sidestepping the potential `null` value.

nightnei replied:
Yep :(

chad1008 commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -257,13 +288,14 @@ class DomainSearch extends Component {
 		}
 
-		return domainManagementList( selectedSiteSlug, currentRoute );
+		return domainManagementList( selectedSiteSlug ?? undefined, currentRoute );
```
Treating a potential `null` value as `undefined` to match the shape of `domainManagementList`

chad1008 commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -416,17 +450,17 @@ class DomainSearch extends Component {
+	( state: IAppState ) => {
 		const site = getSelectedSite( state );
-		const siteId = getSelectedSiteId( state );
+		const siteId = getSelectedSiteId( state ) ?? undefined;
```
Treating potential `null` value as `undefined` to conform to the shape of the selectors we're calling below.

chad1008 commented:
cc @nightnei 

Okay, typing `client/my-sites/domains/domain-search/index.tsx` proved challenging, but it was a good learning experience. You'll absolutely have feedback, there's definitely room for improvement (thanks in advance).

Added inline comments for a couple of spots where I specifically want some advice, but feedback on everything is very welcome.

nightnei commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -435,20 +469,23 @@ export default connect(
-				isSiteOnWooExpress( state, siteId ) ||
-				isSiteOnEcommerce( state, siteId ),
+			isSiteOnFreePlan: !! site && site.plan && isFreePlanProduct( site.plan ),
+			isEcommerceSite: siteId
+				? isSiteOnECommerceTrial( state, siteId ) ||
+				  isSiteOnWooExpress( state, siteId ) ||
+				  isSiteOnEcommerce( state, siteId )
+				: false,
```
We can simplify it w/o hardcodding `false`:
```suggestion
			isEcommerceSite: !! siteId && (
				isSiteOnECommerceTrial( state, siteId ) ||
				  isSiteOnWooExpress( state, siteId ) ||
				  isSiteOnEcommerce( state, siteId )
			),
```

nightnei commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -435,20 +469,23 @@ export default connect(
-				isSiteOnECommerceTrial( state, siteId ) ||
-				isSiteOnWooExpress( state, siteId ) ||
-				isSiteOnEcommerce( state, siteId ),
+			isSiteOnFreePlan: !! site && site.plan && isFreePlanProduct( site.plan ),
```
If we convert site to boolean, hence let's do it for `site.plan`, to have consistency:
```suggestion
			isSiteOnFreePlan: !! site && !! site.plan && isFreePlanProduct( site.plan ),
```

nightnei commented on client/my-sites/domains/domain-search/index.tsx:
```diff
@@ -435,20 +469,23 @@ export default connect(
 			domainsWithPlansOnly: currentUserHasFlag( state, DOMAINS_WITH_PLANS_ONLY ),
-			isSiteUpgradeable: isSiteUpgradeable( state, siteId ),
-			isSiteOnMonthlyPlan: isSiteOnMonthlyPlan( state, siteId ),
+			isSiteUpgradeable: !! siteId && isSiteUpgradeable( state, siteId ),
```
Just thought out loud - I don't like this extra condition, but unfortunately it's indeed the best way, to avoid rewriting all functions like `isSiteUpgradeable` :(

chad1008 committed:
improve typeguard

chad1008 committed:
make `domainTransfer.extra` optional

chad1008 committed:
Apply suggestions from code review

simplify/clean up declarations

Co-authored-by: Volodymyr Makukha <nei.css@gmail.com>

chad1008 commented:
thanks for the feedback and advice @nightnei. All feedback should be addressed, and hopefully good to go

chad1008 committed:
code style fixes

chad1008 committed:
remove old `expect-error`
